In [16]:
!
!pip install -q transformers sentencepiece accelerate

In [2]:
from transformers import pipeline
import torch

In [3]:
# ========== TEXT GENERATION PIPELINES ==========
print("Loading generation pipelines...")

gen_bert = pipeline("text-generation", model="bert-base-uncased")
gen_roberta = pipeline("text-generation", model="roberta-base")
gen_bart = pipeline("text-generation", model="facebook/bart-base")

print("Generation models loaded.\n")


# ========== FILL-MASK PIPELINES ==========
print("Loading fill-mask pipelines...")

mask_bert = pipeline("fill-mask", model="bert-base-uncased")
mask_roberta = pipeline("fill-mask", model="roberta-base")
mask_bart = pipeline("fill-mask", model="facebook/bart-base")

print("Mask models loaded.\n")


# ========== QUESTION ANSWERING PIPELINES ==========
print("Loading QA pipelines...")

qa_bert = pipeline("question-answering", model="bert-base-uncased")
qa_roberta = pipeline("question-answering", model="roberta-base")
qa_bart = pipeline("question-answering", model="facebook/bart-base")

print("QA models loaded ")

Loading generation pipelines...


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

If you want to use `BertLMHeadModel` as a standalone, add `is_decoder=True.`


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

Device set to use cpu


config.json:   0%|          | 0.00/481 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/499M [00:00<?, ?B/s]

If you want to use `RobertaLMHeadModel` as a standalone, add `is_decoder=True.`


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Device set to use cpu


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/558M [00:00<?, ?B/s]

Some weights of BartForCausalLM were not initialized from the model checkpoint at facebook/bart-base and are newly initialized: ['lm_head.weight', 'model.decoder.embed_tokens.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cpu
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


Generation models loaded.

Loading fill-mask pipelines...


Device set to use cpu
Device set to use cpu
Device set to use cpu
Some weights of BertForQuestionAnswering were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['qa_outputs.bias', 'qa_outputs.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Mask models loaded.

Loading QA pipelines...


Device set to use cpu
Some weights of RobertaForQuestionAnswering were not initialized from the model checkpoint at roberta-base and are newly initialized: ['qa_outputs.bias', 'qa_outputs.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Device set to use cpu
Some weights of BartForQuestionAnswering were not initialized from the model checkpoint at facebook/bart-base and are newly initialized: ['qa_outputs.bias', 'qa_outputs.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Device set to use cpu


QA models loaded 


In [4]:
prompt = "The future of Artificial Intelligence is"

In [5]:
print("=== BERT Generation ===")
try:
    output = gen_bert(prompt, max_length=20)
    print(output)
except Exception as e:
    print("BERT failed:", e)

Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Both `max_new_tokens` (=256) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


=== BERT Generation ===
[{'generated_text': 'The future of Artificial Intelligence is................................................................................................................................................................................................................................................................'}]


In [6]:
print("=== RoBERTa Generation ===")
try:
    output = gen_roberta(prompt, max_length=20)
    print(output)
except Exception as e:
    print("RoBERTa failed:", e)

Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Both `max_new_tokens` (=256) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


=== RoBERTa Generation ===
[{'generated_text': 'The future of Artificial Intelligence is'}]


In [7]:
print("=== BART Generation ===")
try:
    output = gen_bart(prompt, max_length=20)
    print(output)
except Exception as e:
    print("BART failed:", e)

Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Both `max_new_tokens` (=256) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


=== BART Generation ===
[{'generated_text': 'The future of Artificial Intelligence isnantss semanticaneouslyuddle ToysCon explicitorientedoriented slashinguddle charred originally kickeduddleuddleoriented slashing kicked���uddle wastefuliterraneanCon charred extremeuddleuddleiterranean explicituddleuddleuddle charred Kasselement charredelementorienteduddleelement charredHero spoileruddle Patreon explicit XIV wastefuluddleuddle explicit wasteful charred wasteful extremeuddle Sheridanelement charred extreme whistleblower Huh frenzyuddle charred explicit SQ Toys ToysTE whistleblower Toys Int extremeHeroHero charredHero Toys Toys Toys extreme extreme Clockwork XIVelement Continental wasteful charred Bhar extremenant charred extremenant extreme extreme Clint Sheridanelement extremeittens extremeHero compound Toys Shen extreme Clint wasteful Griffith blame charredelement extreme satisf extreme satisf Twe accompan wasteful wastefulelement wasteful wasteful wasteful blame extreme extreme extre

### Cell 5 (BERT Generation):
- Errors?: A warning about 'max_length' and 'max_new_tokens' conflicting was observed in the standard error, but no runtime error.
- Repeated words?: Yes, the period character '.' was repeated excessively, making the output mostly unintelligible.
- Random junk?: Yes, the output was effectively random repetitions of a period.
- Does BART behave better?: No, BART's output was significantly worse.

### Cell 6 (RoBERTa Generation):
- Errors?: A similar warning about 'max_length' and 'max_new_tokens' conflict was observed.
- Repeated words?: No. The output was clean but did not extend much beyond the original prompt.
- Random junk?: No, the output was coherent but lacked generative content, essentially echoing the prompt.
- Does BART behave better?: No, BART's output was significantly worse.

### Cell 7 (BART Generation):
Errors?: A similar warning about 'max_length' and 'max_new_tokens' conflict was observed.
- Repeated words?: Yes, many words like 'uddle', 'extreme', 'charred', 'wasteful', and 'whistleblower' were repeated numerous times.
- Random junk?: Yes, the generated text was largely incoherent, appearing as a stream of unrelated and highly repetitive words.
- Does BART behave better?: For this specific text generation task and prompt, BART behaved the worst among the three models, producing highly repetitive and nonsensical output.

In [8]:
masked_text_bert = "The goal of Generative AI is to [MASK] new content."
masked_text_roberta = "The goal of Generative AI is to <mask> new content."
masked_text_bart = "The goal of Generative AI is to <mask> new content."

In [9]:
print("=== BERT Fill Mask ===")
results = mask_bert(masked_text_bert)
for r in results:
    print(r["token_str"], " | score:", round(r["score"], 4))

=== BERT Fill Mask ===
create  | score: 0.5397
generate  | score: 0.1558
produce  | score: 0.0541
develop  | score: 0.0445
add  | score: 0.0176


In [10]:
print("=== RoBERTa Fill Mask ===")
results = mask_roberta(masked_text_roberta)
for r in results:
    print(r["token_str"], " | score:", round(r["score"], 4))

=== RoBERTa Fill Mask ===
 generate  | score: 0.3711
 create  | score: 0.3677
 discover  | score: 0.0835
 find  | score: 0.0213
 provide  | score: 0.0165


In [11]:
print("=== BART Fill Mask ===")
results = mask_bart(masked_text_bart)
for r in results:
    print(r["token_str"], " | score:", round(r["score"], 4))

=== BART Fill Mask ===
 create  | score: 0.0746
 help  | score: 0.0657
 provide  | score: 0.0609
 enable  | score: 0.0359
 improve  | score: 0.0332


In [12]:
context = "Generative AI poses significant risks such as hallucinations, bias, and deepfakes."
question = "What are the risks?"

In [13]:
print("=== BERT QA ===")
result = qa_bert(question=question, context=context)
print(result)

=== BERT QA ===
{'score': 0.009088597260415554, 'start': 72, 'end': 82, 'answer': 'deepfakes.'}


In [14]:
print("=== RoBERTa QA ===")
result = qa_roberta(question=question, context=context)
print(result)

=== RoBERTa QA ===
{'score': 0.00961665902286768, 'start': 0, 'end': 67, 'answer': 'Generative AI poses significant risks such as hallucinations, bias,'}


In [15]:
print("=== BART QA ===")
result = qa_bart(question=question, context=context)
print(result)

=== BART QA ===
{'score': 0.016915911808609962, 'start': 46, 'end': 67, 'answer': 'hallucinations, bias,'}


| Task | Model | Classification (Success/Failure) | Observation (What actually happened?) | Why did this happen? (Architectural Reason) |
| :--- | :--- | :--- | :--- | :--- |
| **Generation** | BERT | *Failure* | *The output consisted mainly of excessive repetitions of the period character '.', making it unintelligible.* | *BERT is primarily an encoder model, not designed for coherent text generation. Its architecture focuses on understanding context, not sequentially generating new text.* |
| | RoBERTa | *Failure* | *The output was coherent but did not extend much beyond the original prompt, effectively echoing it without generating new content.* | *Similar to BERT, RoBERTa is an encoder-only model. While effective for understanding, it's not architecturally suited for open-ended text generation.* |
| | BART | *Failure* | *The generated text was largely incoherent and highly repetitive, with words like 'uddle', 'extreme', 'charred', and 'wasteful' appearing many times.* | *BART is an encoder-decoder model capable of generation, but without fine-tuning on specific generative tasks or with generic prompting, it can produce low-quality, repetitive, or nonsensical output, especially when forced to generate beyond its training distribution for simple prompts.* |
| **Fill-Mask** | BERT | *Success* | *Predicted plausible words like 'create', 'generate', 'produce', 'develop' with high scores.* | *BERT is specifically pre-trained on Masked Language Modeling (MLM), which directly aligns with the fill-mask task.* |
| | RoBERTa | *Success* | *Predicted plausible words such as 'generate', 'create', 'discover', 'find', 'provide' with high scores.* | *RoBERTa also utilizes Masked Language Modeling as a core pre-training objective, making it highly effective for this task.* |
| | BART | *Success* | *Predicted reasonable words like 'create', 'help', 'provide', 'enable', 'improve', though with slightly lower confidence scores than BERT/RoBERTa.* | *BART's pre-training involves reconstructing corrupted text, which includes masked token prediction, allowing it to perform well on fill-mask tasks.* |
| **QA** | BERT | *Partial Success* | *Answered 'deepfakes.' with a low score (0.009). It identified one risk but not all.* | *BERT, when fine-tuned for extractive QA, identifies start and end tokens of an answer span. It found a relevant span, but its confidence was low, and it didn't extract all risks in a single span.* |
| | RoBERTa | *Partial Success* | *Answered 'Generative AI poses significant risks such as hallucinations, bias,' with a low score (0.0096). It extracted a longer but still incomplete span.* | *Similar to BERT, RoBERTa is an extractive QA model. It identified a relevant span, but struggled to capture all distinct risks or a more comprehensive answer, and its confidence was low.* |
| | BART | *Partial Success* | *Answered 'hallucinations, bias,' with a score (0.0169) slightly higher than BERT/RoBERTa, but still missed 'deepfakes'.* | *BART can also perform QA tasks. It successfully identified a relevant portion of the context but, like the other models, did not provide a complete answer encompassing all risks in the given context.* |